### PR TITLE
Creating UrlUtility for IsLocalUrl, and cleaning up UrlHelper

### PR DESCRIPTION
**Pull request by ianhong**
@yishaigalatzer and @rynowak 

As we discussed, I made an internal utility class and refactored UrlHelper accordingly. Please review this PR.

### CODE REVIEW

File: test/Microsoft.AspNet.Mvc.Core.Test/Internal/UrlUtilityTests.cs [hunk 1]
```diff
@@ -0,0 +1,169 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNet.Mvc.Core.Internal;
+using Xunit;
+
+namespace Microsoft.AspNet.Mvc.Core.Test.Internal
+{
```
Comment (ianhong) [inline test/Microsoft.AspNet.Mvc.Core.Test/Internal/UrlUtilityTests.cs]: This is a simply copy of the tests exercising IsLocalUrl from https://github.com/aspnet/Mvc/blob/dev/test/Microsoft.AspNet.Mvc.Core.Test/UrlHelperTest.cs.

**Comment by rynowak** [comment]
Bonus points if you want to address https://github.com/aspnet/Mvc/issues/1889 as part of this change as well. 

I'd suggest something like:

> A URL is considered local if it contains an absolute path, but does not contain a host/authority portion. This also includes URLs using the virtual paths '~/'.

And include a few examples.

File: src/Microsoft.AspNet.Mvc.Core/Internal/UrlUtility.cs [hunk 2]
```diff
@@ -0,0 +1,25 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace Microsoft.AspNet.Mvc.Core.Internal
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Internal/UrlUtility.cs]: Drop the `.Core`

File: test/Microsoft.AspNet.Mvc.Core.Test/Internal/UrlUtilityTests.cs [hunk 3]
```diff
@@ -0,0 +1,169 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNet.Mvc.Core.Internal;
+using Xunit;
+
+namespace Microsoft.AspNet.Mvc.Core.Test.Internal
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Core.Test/Internal/UrlUtilityTests.cs]: `Microsoft.AspNet.Mvc.Internal`

Test code is always in the same namespace as product code.

File: test/Microsoft.AspNet.Mvc.Core.Test/Internal/UrlUtilityTests.cs [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+{
+    public class UrlUtilityTests
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Core.Test/Internal/UrlUtilityTests.cs]: `UrlUtilityTest`

File: src/Microsoft.AspNet.Mvc.Core/Internal/UrlUtility.cs [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+{
+    public static class UrlUtility
+    {
+        /// <summary>
+        /// Returns a value that indicates whether the URL is local.
```
Comment (yishaigalatzer) [inline src/Microsoft.AspNet.Mvc.Core/Internal/UrlUtility.cs]: Make the comment indicate exactly what this does, See #1889

**Comment by rynowak** [comment]
One thing I don't love is the copying of tests. If we end up with duplicates it's easier for them to diverge.

I'd suggest in this case deleting the `UrlHelper` copy of the tests, and replacing it with one simple test that chooses a few complex cases, and has a comment about where to find the thorough tests.

Any other ideas @yishaigalatzer

File: test/Microsoft.AspNet.Mvc.Core.Test/Internal/UrlUtilityTests.cs [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+    {
+		[Theory]
```
Comment (yishaigalatzer) [inline test/Microsoft.AspNet.Mvc.Core.Test/Internal/UrlUtilityTests.cs]: don't use tabs

File: test/Microsoft.AspNet.Mvc.Core.Test/Internal/UrlUtilityTests.cs (hunk 3 above)
Comment (yishaigalatzer) [inline test/Microsoft.AspNet.Mvc.Core.Test/Internal/UrlUtilityTests.cs]: same comment from @rynowak on namespaces here.

File: src/Microsoft.AspNet.Mvc.Core/Internal/UrlUtility.cs (hunk 2 above)
Comment (ianhong) [inline src/Microsoft.AspNet.Mvc.Core/Internal/UrlUtility.cs]: @rynowak I assigned #1889 to myself.

**Comment by ianhong** [comment]
@rynowak and @yishaigalatzer I updated this PR per comments

File: src/Microsoft.AspNet.Mvc.Core/IUrlHelper.cs [hunk 7]
```diff
@@ -33,10 +33,11 @@ public interface IUrlHelper
         string Content(string contentPath);
 
         /// <summary>
-        /// Returns a value that indicates whether the URL is local.
+        /// Returns a value that indicates whether the URL is local. An absoulte URL is considered local
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/IUrlHelper.cs]: If I'm not mistaken, an `absolute URL` **always** has a host/authority section. What we're talking about is a URL with an absolute *path*.

File: src/Microsoft.AspNet.Mvc.Core/IUrlHelper.cs [hunk 8]
```diff
@@ -33,10 +33,11 @@ public interface IUrlHelper
...
+        /// if it does not have a host/authority part. Urls using the virtual paths, '~/' are also local.
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/IUrlHelper.cs]: > Urls using the virtual paths, '~/' are also local

URLs using virtual paths ('~/') are also local.

File: src/Microsoft.AspNet.Mvc.Core/IUrlHelper.cs [hunk 9]
```diff
@@ -33,10 +33,11 @@ public interface IUrlHelper
...
         /// </summary>
         /// <param name="url">The URL.</param>
-        /// <returns>true if the URL is local; otherwise, false.</returns>
+        /// <returns><c>true</c> if the URL is local; otherwise, <c>false</c>.</returns>
         bool IsLocalUrl(string url);
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/IUrlHelper.cs]: Add some examples to the docs of local and non-local using `<example>`

**Comment by rynowak** [comment]
:watch: docs need a little more precision, and some examples.

**Comment by ianhong** [comment]
@rynowak Updated this PR per comments

File: src/Microsoft.AspNet.Mvc.Core/UrlHelper.cs [hunk 10]
```diff
@@ -71,14 +72,7 @@ public UrlHelper(IScopedInstance<ActionContext> contextAccessor, IActionSelector
...
         {
-            return
-                !string.IsNullOrEmpty(url) &&
-
-                // Allows "/" or "/foo" but not "//" or "/\".
-                ((url[0] == '/' && (url.Length == 1 || (url[1] != '/' && url[1] != '\\'))) ||
-
-                // Allows "~/" or "~/foo".
-                (url.Length > 1 && url[0] == '~' && url[1] == '/'));
+			return UrlUtility.IsLocalUrl(url);
```
Comment (yishaigalatzer) [inline src/Microsoft.AspNet.Mvc.Core/UrlHelper.cs]: format the code, and stop using tabs for indentation

File: test/Microsoft.AspNet.Mvc.Core.Test/Internal/UrlUtilityTest.cs [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+
+using Xunit;
+
+namespace Microsoft.AspNet.Mvc.Internal
+{
+    public class UrlUtilityTest
+    {
+        [Theory]
+        [InlineData(null)]
+        [InlineData("")]
```
Comment (yishaigalatzer) [inline test/Microsoft.AspNet.Mvc.Core.Test/Internal/UrlUtilityTest.cs]: nit: Add " "

File: test/Microsoft.AspNet.Mvc.Core.Test/Internal/UrlUtilityTest.cs [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+            var result = UrlUtility.IsLocalUrl(url);
+
+            // Assert
+            Assert.False(result);
+        }
+
+        [Theory]
+        [InlineData("http://www.mysite.com/appDir/foo.html")]
+        [InlineData("http://WWW.MYSITE.COM")]
+        public void IsLocalUrl_RejectsUrlsOnTheSameHost(string url)
```
Comment (yishaigalatzer) [inline test/Microsoft.AspNet.Mvc.Core.Test/Internal/UrlUtilityTest.cs]: what do you mean by same host? how is www.mysite.com special?

**Comment by yishaigalatzer** [comment]
Address comments and :shipit: when @rynowak is happy

File: test/Microsoft.AspNet.Mvc.Core.Test/Internal/UrlUtilityTest.cs (hunk 12 above)
Comment (ianhong) [inline test/Microsoft.AspNet.Mvc.Core.Test/Internal/UrlUtilityTest.cs]: @yishaigalatzer I think this test name was made when this method is an instance method. I'll remove this test and the IsLocalUrl_RejectsUrlsOnTheSameHostButDifferentScheme test, and then combine them to an appropriate test.

File: src/Microsoft.AspNet.Mvc.Core/Internal/UrlUtility.cs [hunk 13]
```diff
@@ -0,0 +1,39 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace Microsoft.AspNet.Mvc.Internal
+{
+    public static class UrlUtility
+    {
+        /// <summary>
+        /// Returns a value that indicates whether the URL is local. An URL with an absolulte path is considered local
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Internal/UrlUtility.cs]: `absolulte` -> `absolute`

File: test/Microsoft.AspNet.Mvc.Core.Test/Internal/UrlUtilityTest.cs (hunk 12 above)
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Core.Test/Internal/UrlUtilityTest.cs]: @yishaigalatzer - the naming of this is historical cruft. These tests and code were copied almost verbatim for reasons.

File: src/Microsoft.AspNet.Mvc.Core/Internal/UrlUtility.cs (hunk 13 above)
Comment (ianhong) [inline src/Microsoft.AspNet.Mvc.Core/Internal/UrlUtility.cs]: oops. I thin I should install a spell checker extension..

**Comment by ianhong** [comment]
@rynowak Updated the tests per our discussion and fixed the typo in the doc.

File: test/Microsoft.AspNet.Mvc.Core.Test/Internal/UrlUtilityTest.cs [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+        [InlineData("/\\foo", false)]
+        public void IsLocalUrl_ReturnsExpectedResults(string url, bool expectedResult)
+        {
+            // Arrange & Act
+            var result = UrlUtility.IsLocalUrl(url);
+
+            // Assert
+            Assert.Equal(expectedResult, result);
+        }
+    }
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Core.Test/Internal/UrlUtilityTest.cs]: There's no real need for this file as both are covered directly.

File: test/Microsoft.AspNet.Mvc.Core.Test/UrlHelperTest.cs [hunk 15]
```diff
@@ -61,6 +62,7 @@ public class UrlHelperTest
         [Theory]
         [InlineData(null)]
         [InlineData("")]
+        [InlineData(" ")]
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Core.Test/UrlHelperTest.cs]: Add a comment somewhere obvious about **why** both code paths are covered by these tests

**Comment by rynowak** [comment]
:shipit: after fixing minor comments

**Pull request merged**